Projeto 7

AlanDouglasASouza opened the pull request:
Hi guys, this PR implements the routine of solving a Todo in the project. For this new feature:

• We have listed the option to resolve a Todo.
• We show the list of numbered tasks.
• We allow resolution of the chosen Todo.
• When listing all Todo's again, Todo's that have already been resolved are displayed in a different style.

[Project 7](https://github.com/terramagna/onboarding-rust/blob/main/essenciais/2-intermediario/09-projeto-7/README.md).

AlanDouglasASouza committed:
chore: add property resolved in Todo struct

Add 'resolved' in Todo, and modify method ask_for_new_todo of trait UserInterface for return Todo whit 'resolved'.

AlanDouglasASouza committed:
refac: add resolve in Todo and resolve_one_todo in TodoStorage trait

• Add resolve variant to the UserComand enum, in preparation for the resolve feature.
• Add resolve method to struct Todo.
• Add the resolve_one_todo method to Todos, keeping the change of a Resolved Todo.

AlanDouglasASouza committed:
feat: add resolve Todo's

• Adds the solve a todo routine.
• Add methods in TodoCli and in the UserInterface trait that allow you to resolve Todo's, and highlight the ones that were resolved.

luisholanda commented on src/cli.rs:
```diff
@@ -112,10 +113,33 @@ impl TodoCli {
+                    self.todo_storage.resolve_one_todo(key);
+                    self.user_interface
+                        .write_feedback("✅ TODO resolvido com sucesso! ✅")?;
+                    break;
```
Dado que essa é a única forma de sair do loop e não há mais nenhuma operação além disso, o ideal aqui seria retornar direto ao invés de usar um `break`, deixa o código mais claro e fácil de ler.

AlanDouglasASouza replied:
OK! Fica mais idiomático mesmo!

luisholanda commented on src/terminal.rs:
```diff
@@ -93,15 +96,20 @@ impl UserInterface for Terminal {
-        let todo_style = style(&*todo.message).yellow().italic();
-        let todo_msg = msg_initial.to_owned() + &todo_style.to_string();
+        let todo_msg = match todo.resolved {
+            false => msg_initial.to_owned() + &style(&*todo.message).yellow().italic().to_string(),
+            true => "✅: ".to_owned() + &style(&*todo.message).yellow().italic().dim().to_string(),
```
Eu usaria `format!` aqui, é bem mais claro.

Em geral, o uso de `+` com strings em Rust é _bem raro_.

AlanDouglasASouza replied:
Nice! O macro `format!` é vida. :love_you_gesture:

luisholanda commented on src/todo.rs:
```diff
@@ -2,6 +2,16 @@ use std::fmt::{Display, Formatter, Result};
+}
+
+impl Todo {
+    pub fn resolve(&self) -> Self {
+        Self {
+            message: self.message.clone(),
+            resolved: true,
+        }
```
Hmmm, acho que aqui ficaria mais idiomático se a atualização fosse feita in-place.

AlanDouglasASouza replied:
Eu ponderei nessa hora mesmo, eu vou abstrair esse método e fazer a atualização onde é chamado, não tenho certeza se seria assim a sua sugestão. WDYM? :thinking:

AlanDouglasASouza committed:
refac: remove resolve method from Todo

Remove the resolve method from Todo and add fomart! for show_todo of UserInterface trait.

luisholanda commented on src/terminal.rs:
```diff
@@ -93,15 +96,20 @@ impl UserInterface for Terminal {
         let new_todo = self.input()?;
 
-        Ok(Todo { message: new_todo })
+        Ok(Todo {
+            message: new_todo,
+            resolved: false,
+        })
```
Acho que o mais adequado aqui seria termos um `Todo::new(new_todo)`.

AlanDouglasASouza replied:
Blz! Adicionei um `Todo::new(todo)`. :rocket:

luisholanda commented on src/todos.rs:
```diff
@@ -53,4 +54,17 @@ impl TodoStorage for Todos {
     }
+
+    fn resolve_one_todo(&mut self, key: u32) -> bool {
+        match self.get_one_todo(key) {
+            Some(todo) => self.update(
+                key,
+                Todo {
+                    message: todo.message.clone(),
+                    resolved: true,
+                },
+            ),
+            None => false,
+        }
```
Não acho que é necessario tentar reusar os métodos já existentes aqui.

```suggestion
        let Some(todo) = self.todo_collection.get_mut(key) else {
            return false;
        };

        todo.resolved = true;

        true
```

AlanDouglasASouza replied:
Ok! Usar o `get_mut` aqui ficou bem interessante por ele retornar um `&mut Todo`.

luisholanda reviewed (commented):
Algumas sugestões de melhoria, mas já está bem legal :)

AlanDouglasASouza committed:
refac: change resolve_one_todo

Change resolve_one_todo for get_mut from BTreeMap.
Add the new method to Todo.
